---
title: "Simulation of the evolution of gene expression along the Capsella phylogeny"
author: "Sylvain  Glemin"
date: "08/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
if (!require("sde"))   install.packages("sde", dependencies = TRUE)
if (!require("ape"))   install.packages("ape", dependencies = TRUE)
library(sde)
library(ape)
set.seed(123)
```

# Model

The aim of this notebook is to generate simulated datasets of gene expression evolution along the Capsella phylogeny under a neutral model. The branch length leading to Co is normalized to $1$ and the length of the branch leading to the ancestor of Cg and Cr is $1-x$. In what follows, $x=0.05$ is used but it can be modified. It roughly corresponds to assuming 1 Myr divernce between C. grandiflora and C. orientalis and about 50 Kyr for tyhe origin of C. rubella.

```{r}
caps_tree <- read.tree(text = "(Co:1,(Cg:0.1,Cr:0.1):0.9);")
plot(caps_tree)
edgelabels(c("1","1-x","x","x"))
```


The evolution of gene expression can be simulated as an Ornstein-Ulhenbeck (OU) process running on the tree. A OU process can be described by the following stochastic differential equation:
\begin{equation}
  dX = \alpha(\theta - X) + \sigma dB
\end{equation}
where $\theta$ is the optimum trait value, $\alpha$ the strength of the force moving the trait to the optimum (~selection), $dB$ a Wiener process with variance $\sigma^2$ (~drift).
For simplicity we assume that the ancestral value is $X_0 = 0$ and that the optimum value is also $\theta = 0$. As we are interested in simulating the null hypothesis, we consider constant parameters along the phylogeny. When $\alpha$ tends towards 0 it is equivalent to a Brownian motion, so expression evolving under pure drift. If $\alpha >0$, it corresponds to uniform stabilizing selection along the phylogeny.

The following function generates a simulated dataset.
Default values are given. Note that the parameter $\alpha$ cannot be set exactly to 0 so a very low value must be used to simulate a brownian motion. nsim genes are simultaneously simulated (default 20000).

```{r}
simul <- function(alpha=10^(-10), sigma=1, x=0.05, nsim = 20000){
  Co <- rcOU(n=nsim, Dt=1, x0=0, theta=c(0,alpha,sigma))
  CgCr <- rcOU(n=nsim, Dt=1-x, x0=0, theta=c(0,alpha,sigma))
  Cg <- rcOU(n=nsim, Dt=x, x0=CgCr, theta=c(0,alpha,sigma))
  Cr <- rcOU(n=nsim, Dt=x, x0=CgCr, theta=c(0,alpha,sigma))
  return(list("Co"=Co,"Cg"=Cg,"Cr"=Cr))
}
```


The following function generates a plot of the distribution of the two indices
```{r}
# Function to calculate the Milesi's indices
Dstat <- function(x,y)  (x - y)/max(x,y)

plot_indices <- function(dataset) {
  distCoCr <- abs(dataset$Co-dataset$Cr)
  distCoCg <- abs(dataset$Co-dataset$Cg)
  distCrCg <- abs(dataset$Cr-dataset$Cg)
  DCr <- mapply(Dstat, x=distCoCr,y=distCoCg)
  DCo <- mapply(Dstat, x=distCoCr,y=distCrCg)
  layout(matrix(c(1,2),1,2))
  hist(DCr,main=paste("DCr (mean = ",round(mean(DCr),2),")",sep=""),breaks=50)
  hist(DCo,main=paste("DCo (mean = ",round(mean(DCo),2),")",sep=""),breaks=50)
}
```


# Examples

In what follows the distribution of the two indices are plotted. However, only DCr is commented because it has a simpler interpretation and direct prediction can be made as its null distribution is symmetrical around 0.

## Browian model

All genes are supposed to have the same parameters

```{r}
datanull <- simul()
plot_indices(datanull)

datanull <- simul(sigma = 10)
plot_indices(datanull)

datanull <- simul(sigma = 0.1)
plot_indices(datanull)
```
It is almost not affected by the magntiude of drift ($\sigma$)

We can also simulate each gene with its own $\sigma$.

```{r}
tabsigma <- rlnorm(n = 20000,meanlog = 0,sdlog = 1) # Example given with sigma drawn in a log-normal distribution. Can be anything, including emprical values
sim <- sapply(tabsigma,function(x) simul(sigma = x,nsim = 1))
datanull2 <- list("Co"=unlist(sim[1,]),"Cg"=unlist(sim[2,]),"Cr"=unlist(sim[3,]))
plot_indices(datanull2)
```

This does not change the distribution.


## OU process

```{r}
dataOU <- simul(alpha = 0.1)
plot_indices(dataOU)

dataOU <- simul(alpha = 1)
plot_indices(dataOU)

dataOU <- simul(alpha = 5)
plot_indices(dataOU)

dataOU <- simul(alpha = 10)
plot_indices(dataOU)

dataOU <- simul(alpha = 50)
plot_indices(dataOU)
```


Increasing $\alpha$ changes the global shape but doesn't not change the asymmetry.

It is also possible to assume that each gene as it's own $\sigma$ and $\alpha$.

```{r}
tabsigma <- rlnorm(n = 20000,meanlog = 0,sdlog = 1)
tabalpha <- rlnorm(n = 20000,meanlog = 2,sdlog = 1)
sim <- mapply(function(x,y) simul(sigma = x,alpha = y,nsim = 1), x = tabsigma, y=tabalpha)
dataOU2 <- list("Co"=unlist(sim[1,]),"Cg"=unlist(sim[2,]),"Cr"=unlist(sim[3,]))
plot_indices(dataOU2)
```


